core/base/index_set.cpp

```diff
@@ -0,0 +1,162 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/index_set.hpp>
+
+
+#include <algorithm>
+#include <iostream>
+#include <mutex>
+#include <vector>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+#include "core/base/index_set_kernels.hpp"
+
+
+namespace gko {
+namespace index_set {
+
+
+GKO_REGISTER_OPERATION(to_global_indices, index_set::to_global_indices);
+GKO_REGISTER_OPERATION(populate_subsets, index_set::populate_subsets);
+GKO_REGISTER_OPERATION(global_to_local, index_set::global_to_local);
+GKO_REGISTER_OPERATION(local_to_global, index_set::local_to_global);
+
+
+}  // namespace index_set
+
+
+template <typename IndexType>
+void IndexSet<IndexType>::populate_subsets(const gko::Array<IndexType>& indices,
+                                           const bool is_sorted)
+{
+    auto exec = this->get_executor();
+    this->num_stored_indices_ = indices.get_num_elems();
+    exec->run(index_set::make_populate_subsets(
+        this->index_space_size_, &indices, &this->subsets_begin_,
+        &this->subsets_end_, &this->superset_cumulative_indices_, is_sorted));
```

Comment:
nit: why pass arrays as pointers? passing them as references saves & and *

Reply:
Just a personal preference from my side. It uniformizes reading the kernel when passing plain pointers, I can use `->` everywhere, instead of switching between `->` and `.`